### DISCUSSION TITLE
WeAct F411 'blackpill' can't upload .hex file

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#16478] [STM32 / Pyboard]
Title should be:
Can not run the .hex file for V3.1 and 8MB SPI flash
v1.25.0-preview.160.gc73204128
using STM32Cube Programmer and get the following:
```
dave@davef:~$ rsusb
Using buffer-size of 30
Connecting to /dev/ttyUSB0 (buffer-size 30)...
Trying to connect to REPL  connected
Retrieving sysname ... b'FATAL: uncaught exception 200039d0\r\nOSError: 16\r\n\nFATAL ERROR:\n'
Traceback (most recent call last):
  File "/home/dave/.local/lib/python3.10/site-packages/rshell/main.py", line 1381, in connect
    ip_address = socket.gethostbyname(port)
socket.gaierror: [Errno -2] Name or service not known
```

**Comment by shariltumin** [comment]
The WEACT_F411CEU6 board can be flashed using DFU.

To put the board into DFU mode, disconnect USB, connect `A10` to 3V3. 

```
Press and hold `BOOT0`.
Insert USB cabel.
Wait 0.5 seconds.
Release `BOOT0`.
```

OR

```
Insert USB cabel
Press and hold `BOOT0`. 
Press and release `RESET`. 
Wait 0.5 seconds.
Release `BOOT0`.
``` 

Then 
```
$ lsusb
...
Bus 001 Device 010: ID 0483:df11 STMicroelectronics STM Device in DFU Mode
...
```

You should see your board as "STM Device in DFU Mode". You are now ready to flash the MP firmware using ```dfu```.

```
$ dfu -u firmware.dfu
File: firmware.dfu
    DfuSe v1, image size: 321221, targets: 1
    Target 0, alt setting: 0, name: "ST...", size: 320936, elements: 2
      0, address: 0x08000000, size: 14696
      1, address: 0x08020000, size: 306224
    usb: 0483:df11, device: 0x0000, dfu: 0x011a, UFD, 16, 0x9ea6ff30
Writing memory...
0x08000000   14696 [=========================] 100%
0x08020000  306224 [=========================] 100%
Exiting DFU...
```

Disconnect USB, remove jumper between A10 and 3v3. Enjoy!
By the way, Merry Christmas and a Happy New Year.

**Reply by davefes** [reply]
I have been using .hex files for the last few years as there were problems with BlackPills and DFU mode.  I will try DFU.

Season's greetings as well.

**Reply by scruss** [reply]
I found the timing of the *wait ½ second, release BOOT0* extremely sensitive on my F411 boards. I can get it to work maybe one time out of ten

**Reply by shariltumin** [reply]
This also works, short A10 and 3V3. Some say you need a pullup resistor, but straight short works for me. Press and hold BOOT0. Plug board into USB. Release BOOT0 when the board is securely in place. I assume A10 is not connected to any component in your setup.

**Reply by davefes** [reply]
Firstly, my title is incorrect ... it should read `Can't run the provided .hex file on the device`.  How do you edit the title?

There were many attempts at using DFU on this board here:
https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython/issues?q=is%3Aissue+is%3Aclosed
The conclusion was that poor tolerance on the 25MHz crystal resulted in many hacky attempts at getting DFU mode to work.   WeAct changed the 25MHz to 8MHz to eliminate (reduce) the problems of getting into DFU mode.  I haven't bought any 8MHz boards.

As using STM32Cube Programmer and .hex files worked with other distros including the initial work on the firmware that is now part of the `preview releases` I thought that using that workflow would still be the most convenient.

**Reply by shariltumin** [reply]
When you ran ```rsusb```, did you actually run rshell? It is strange that the program tries to connect to /dev/ttyUSB0.

On my machine, rshell gave me:
```
$ rshell -p /dev/ttyACM0 
Using buffer-size of 512
Connecting to /dev/ttyACM0 (buffer-size 512)...
Trying to connect to REPL  connected
Retrieving sysname ... stm32f411
Testing if ubinascii.unhexlify exists ... Y
Retrieving root directories ... /flash/
Setting time ... Dec 25, 2024 20:31:55
Evaluating board_name ... pyboard
Retrieving time epoch ... Jan 01, 2000
Welcome to rshell. Use Control-D (or the exit command) to exit rshell.
```

I prefer to use mpremote, I do not know how to use rshell.
```
$ mpremote repl /dev/ttyACM0 
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] or Ctrl-x to exit this shell

>>> 
>>> import sys
>>> sys.implementation
(name='micropython', version=(1, 25, 0, 'preview'), _machine='WeAct F411CEU6 8MB with STM32F411CE', _mpy=7942)

```

You can try mpremote. See if the MP firmware gives you REPL.

**Reply by shariltumin** [reply]
You may find the information useful. Especially if you have not seen it before.
https://docs.qmk.fm/platformdev_blackpill_f4x1#bootloader-issues

**Reply by scruss** [reply]
> This also works, short A10 and 3V3. 

This works (with the 8 MHz ones, at least). I'd been using mcauser's [Flashing via DFU](https://github.com/mcauser/WEACT_F411CEU6?tab=readme-ov-file#flashing-via-dfu) instructions of shorting A9 to A10

**Reply by shariltumin** [reply]
All my boards are 25MHz type. I used to use the A9+A10 methods, which gave me a 10% success rate. With A10+GND I get 90% success rate so far.

**Reply by davefes** [reply]
```
alias rsusb='rshell -p /dev/ttyUSB0 --baud=115200 --editor nano --buffer-size=30 -a'
```
Other .hex work with this.  
```
cd /dev`
ls tty*
ttyUSB0
```
no ttyACM0

Thanks for the hint however I suspect that there is something wrong in the header or ?

**Reply by davefes** [reply]
I have seen similar "fixes" to the link you posted.  I will attempt DFU if this can't be sorted.  I will raise an issue.

8MHz are suppose to be much easier to get into DFU mode.

**Reply by davefes** [reply]
OK, giving DFU mode another go.

For Ubuntu the program appears to be `dfu-utils`.  Also, I put a 10K pull-up on A10 as suggested in the link.

The BlackPill appears to go into DFU mode but seems there are 4 devices connected:

```
dave@davef:~/Downloads$ dfu-util -U WEACT_F411_BLACKPILL-V31_FLASH_8M-20241223-v1.25.0-preview.160.gc73204128.dfu
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: More than one DFU capable USB device found! Try `--list' and specify the serial number or disconnect all but one device

dave@davef:~/Downloads$ dfu-util --list
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Found DFU: [0483:df11] ver=2200, devnum=6, cfg=1, intf=0, path="1-4", alt=3, name="@Device Feature/0xFFFF0000/01*004 e", serial="34A938743233"
Found DFU: [0483:df11] ver=2200, devnum=6, cfg=1, intf=0, path="1-4", alt=2, name="@OTP Memory /0x1FFF7800/01*512 e,01*016 e", serial="34A938743233"
Found DFU: [0483:df11] ver=2200, devnum=6, cfg=1, intf=0, path="1-4", alt=1, name="@Option Bytes  /0x1FFFC000/01*016 e", serial="34A938743233"
Found DFU: [0483:df11] ver=2200, devnum=6, cfg=1, intf=0, path="1-4", alt=0, name="@Internal Flash  /0x08000000/04*016Kg,01*064Kg,03*128Kg", serial="34A938743233"
```
I checked by removing the BlackPill and all 4 devices disappear.

Any suggestions how to work out which device I want?  I'd guess `Internal Flash` but ...

**Reply by davefes** [reply]
Went back to a command I used a year or two ago:
`sudo dfu-util --alt 0 -D WEACT_F411_BLACKPILL-V31_FLASH_8M-20241223-v1.25.0-preview.160.gc73204128.dfu`
and the file appears to upload correctly ... but when run I get the same error I got with the hex file.

**Comment by davefes** [comment]
The board has a faulty 8MB flash, I could program it with the default (non-external flash variant).
